### PR TITLE
Avoid changing state when rehydrated messages same as incoming ones.

### PR DESCRIPTION
**Pull request by Sam1301**
This prevents some re-renders of Message related components during startup.

Added a test to cover this case.

### CODE REVIEW

File: src/chat/__tests__/chatReducers-test.js [hunk 1]
```diff
@@ -656,5 +656,25 @@ describe('chatReducers', () => {
...
+      const action = deepFreeze({
+        type: MESSAGE_FETCH_SUCCESS,
+        narrow: [],
+        messages: [{ id: 2, timestamp: 4 }, { id: 3, timestamp: 5 }],
+        replaceExisting: true,
+      });
+
+      const newState = chatReducers(initialState, action);
+
+      expect(...newState.messages[homeNarrowStr]).toBe(...commonMessages);
```
Comment (borisyankov) [inline src/chat/__tests__/chatReducers-test.js]: Why the '...' here?

File: src/chat/__tests__/chatReducers-test.js [hunk 2]
```diff
@@ -656,5 +656,25 @@ describe('chatReducers', () => {
 
       expect(newState).toEqual(expectedState);
     });
+
+    test('when replaceExisting is true, common messages are not replaced', () => {
```
Comment (borisyankov) [inline src/chat/__tests__/chatReducers-test.js]: It is possible that the message has been changed, even though the id is the same the message contents will not be.
I guess we need to handle that.

File: src/chat/chatReducers.js [hunk 3]
```diff
@@ -56,7 +56,7 @@ export default (state: ChatState = initialState, action: Action) => {
       const messages = state.messages[key] || [];
 
       const newMessages = action.replaceExisting
-        ? action.messages
+        ? action.messages.map(item => messages.find(msg => msg.id === item.id) || item)
```
Comment (borisyankov) [inline src/chat/chatReducers.js]: Have a look at the `getStreamsById` selector.
We were considering `getMessagesById` which we haven't have implemented yet. That might be worth doing for perf reasons, for these iterations (there are similar iterations elsewhere).

Consider this here, but also can be implemented as another PR, don't let this slow you down for this one.

Comment (Sam1301) [inline src/chat/chatReducers.js]: this is a great suggestion, added the `getMessagesById` selector 👍

File: src/chat/__tests__/chatReducers-test.js (hunk 1 above)
Comment (Sam1301) [inline src/chat/__tests__/chatReducers-test.js]: I wanted to perform `===` over the individual messages (not array).

File: src/chat/__tests__/chatReducers-test.js (hunk 2 above)
Comment (Sam1301) [inline src/chat/__tests__/chatReducers-test.js]: I missed this... updated to perform a deep equal.

Now one thing to consider is if the win we get from not re-rendering better than the cost for performing a deep equal. For that I have timed the startup for both cases, and gonna continue this discuss though chat.

File: src/chat/__tests__/chatReducers-test.js (hunk 1 above)
Comment (borisyankov) [inline src/chat/__tests__/chatReducers-test.js]: Ah, makes sense... but there is a better way.
Use `toEqual` which is a deep comparison. `toBe` is a referential comparison.

Comment (Sam1301) [inline src/chat/__tests__/chatReducers-test.js]: Sure, before making this change just wanna add this
Here, I intended to use a referential comparison over messages
if we use `toEqual`, even the following will pass the test case
```javascript
      const differentArr = [{ id: 2, timestamp: 4 }, { id: 3, timestamp: 5 }];

      expect(...newState.messages[homeNarrowStr]).toEqual(...differentArr);
```

and if we use `toBe`, the above test case would fail making sure that we maintain referential equality over common messages (and thus prevent needless re-renders)
